### PR TITLE
[3D] Added option to Game View Editor SpawnContextMenu

### PR DESCRIPTION
**Pull request by sili3011**
to save positional and rotational changes of SpawnInstances to database.

**Use case**
Making changes to position and/or rotation of creatures or gameobjects in the Game View Editor and saving those changes to the database.

Previous workflow:
Selecting spawn in game view, modifying it, opening the table with double click on spawn and typing changed values by hand into the form.

image

Workflow is tedious and error prone.

New workflow:
Selecting spawn in game view, modifying it, right click anywhere and select option "Update values".

image

### CODE REVIEW

File: Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs [hunk 1]
```diff
@@ -1,23 +1,111 @@
...
+        ITableDefinitionProvider definitionProvider,
+        IMessageBoxService messageBoxService)
     {
         this.spawnSelectionService = spawnSelectionService;
+        this.mySqlExecutor = mySqlExecutor;
+        this.definitionProvider = definitionProvider;
+        this.messageBoxService = messageBoxService;
 
         CopyGuidCommand = new DelegateCommand<SpawnInstance>(inst => clipboardService.SetText(inst.Guid.ToString()));
+        CopyPositionXCommand = new DelegateCommand<SpawnInstance>(inst => clipboardService.SetText(inst.WorldObject!.Position.X.ToString()));
```
Comment (BAndysc) [inline Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs:42]: I wonder, is it worth to have this as three separate menu items instead of one "Copy Position" which would copy comma separated values? `x, y, z` Just wondering, if it is better to have them separate I am fine I guess.

File: Rendering/WDE.MapSpawns/Rendering/SpawnGizmo.cs [hunk 2]
```diff
@@ -74,7 +74,7 @@ public override Quaternion GetRotation(SpawnInstance item)
         if (item is CreatureSpawnInstance cr)
             return Quaternion.CreateFromAxisAngle(Vectors.Up, cr.Creature!.Orientation);
         else if (item is GameObjectSpawnInstance go)
-            return go.GameObject!.Rotation;
+            return Quaternion.CreateFromAxisAngle(Vectors.Up, go.GameObject!.Orientation);
```
Comment (BAndysc) [inline Rendering/WDE.MapSpawns/Rendering/SpawnGizmo.cs:77]: gameobjects can be rotated in all axes, why this change?

File: Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs [hunk 3]
```diff
@@ -1,23 +1,111 @@
...
+    private ICommand CopyOrientationCommand { get; }
+    private ICommand UpdateValuesCommand { get; }
 
-    public SpawnContextMenu(ISpawnSelectionService spawnSelectionService,
-        IClipboardService clipboardService)
+    public SpawnContextMenu(
+        ISpawnSelectionService spawnSelectionService,
+        IClipboardService clipboardService,
+        ITableEditorPickerService tableEditorPickerService,
+        IDatabaseQueryExecutor mySqlExecutor,
```
Comment (BAndysc) [inline Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs:32]: I think here I'd use `IMySqlExecutor` instead of `IDatabaseQueryExecutor`, those are bad names I can see now...
`IMySqlExecutor` - executes queries on the world database
`IMySqlHotfixExecutor` - executes queries on the hotfixes database
`IDatabaseQueryExecutor` - has additional parameter which determines if the query should be executed in the world database or hotfixes database.

I can see why you chose the `IDatabaseQueryExecutor` here, however the intention was to keep  `IDatabaseQueryExecutor` only for the `WDE.DatabaseEditors` project. While database structure varies between different wow core, I think we can safely assume that creatures/gameobjects tables will be always in the world database.

File: Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs [hunk 4]
```diff
@@ -1,23 +1,111 @@
...
+        UpdateValuesCommand = new DelegateCommand<SpawnInstance>(async (inst) =>
+        {
+            var trans = Queries.BeginTransaction();
+
+            CreatureSpawnInstance? creature = inst as CreatureSpawnInstance;
+            GameObjectSpawnInstance? go = inst as GameObjectSpawnInstance;
+
+            var typeString = creature != null ? "creature" : "gameobject";
+
+            trans.Table(typeString).InsertIgnore(new { Guid = inst.Guid });
```
Comment (BAndysc) [inline Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs:68]: I think it is not a good idea to have an insert query here, this could insert an invalid row to the database (no creature entry, no phasemask, no spawnmask), beside you can only edit existing objects now, so probably better to remove it for now.

File: Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs [hunk 5]
```diff
@@ -1,23 +1,111 @@
...
+                clipboardService.SetText(creature.Creature!.Orientation.ToString()); 
+            }
+            else if (go != null) 
+            {
+                clipboardService.SetText(go.GameObject!.Orientation.ToString());
+            }
+        });
+        UpdateValuesCommand = new DelegateCommand<SpawnInstance>(async (inst) =>
+        {
+            var trans = Queries.BeginTransaction();
```
Comment (BAndysc) [inline Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs:61]: Very good job on finding and using the syntax for writing SQL queries! :D

However, as you said, compatibility with different cores is a question. In case of creature/gameobject position I think the db structure is the same, but... there is a way of writing core-independent queries in the editor.

 There is no code for generating core-independent update position query, but let me show how it works for other query - updating quest chaining. 

```
IQueryGenerator<QuestChainDiff> queryGenerator; // get via constructor
var query = queryGenerator.Update(new QuestChainDiff() {
     Id = 1235,
     PrevQuestId = 1,
     NextQuestId = 2
});
mysqlExecutor.Execute(query);
```

how it works underneath: there are two implementations for `IUpdateQueryProvider<QuestChainDiff>`: `TrinityQuestQueryProvider` `MangosQuestQueryProvider`, which generate query like you generate it here.


So it would be great if you could create  `CreatureSpawnModelDiff` struct with an entry, guid, positions, orientations, create a `class CreatureUpdateQueryProvider : IUpdateQueryProvider<CreatureSpawnModelDiff>`, and in SpawnContextMenu use `IQueryGenerator<CreatureSpawnModelDiff>`. (note: there is already a `CreatureSpawnModelEssentials` struct which is used by query generator for insert and delete, but it is not really suitable for updating as its fields are not nullable

This way we can write integrations test for query generators (look at `DatabaseTester/QueryGeneratorTester.cs` for existing tests).

**Review by BAndysc — CHANGES_REQUESTED** [review]
Thank you for the contribution! I can see you grasped how the code works really fast! I've added few comments, it would be great if you could take a look at them.

File: Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs (hunk 5 above)
Comment (sili3011) [reply Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs:61]: I actually got the general gist of how to do it from ```TrinityQuestQueryProvider``` but didnt think that far that this of course also provides the opportunity to do a core-independant implementation. Thanks for pointing that out, I ll try adding that!

File: Rendering/WDE.MapSpawns/Rendering/SpawnGizmo.cs (hunk 2 above)
Comment (sili3011) [reply Rendering/WDE.MapSpawns/Rendering/SpawnGizmo.cs:77]: Yeah, this is a weird one and probably needs some testing. When i spawned my first few gameobjects and tried rotating them I didnt get the same results in game as in the editor. Then i found out that gameobject (just like creature) has also orientation and started using that, despite of course losing two axis. But I know what you mean, from a logic stand point this doesnt make sense, but it works. Will test some more and go back to inital implementation once its working how its supposed to.

File: Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs (hunk 1 above)
Comment (sili3011) [reply Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs:42]: Before implementing the "Update values" option i just had a copy option for all the values that needed updating, so I could just click "Copy foo", open DB Table Editor, and paste the copied value into the column foo. I deleted most of the options as those got obsolete because of the new "Update values", but figured the positions could still be useful to keep. I think you are right, combining them, now that there is the new save option, is probably the way to go.

File: Rendering/WDE.MapSpawns/Rendering/SpawnGizmo.cs (hunk 2 above)
Comment (sili3011) [reply Rendering/WDE.MapSpawns/Rendering/SpawnGizmo.cs:77]: So there is something different in how the GameView renders some things and I think this i due to rotation parameters, but I cant put my finger where and why and I also kinda dont care enough :D this only happens when hand modifying some rotation values in isolation.

Game View Editor:

image

In game:

image

As you can see gameobjects can get stretched/squished, game seems to make sure that no matter the rotational values the gameobject always has the right proportions.

As this has nothing to do with my "fix" I reverted back to your initial implementation

File: Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs (hunk 5 above)
Comment (sili3011) [reply Rendering/WDE.MapSpawns/Rendering/SpawnContextMenu.cs]: Done! Hope its to your liking. As far as I can see cmangos uses the same column names as trinity [creature](https://github.com/cmangos/issues/wiki/creature) [gameobject](https://github.com/cmangos/issues/wiki/gameobject).

**Review by BAndysc — APPROVED** [review]
Sorry for the delay and thanks for the PR, merged! A small thing, I will split PositionRotationDiff into CreatureDiff and GameobjectDiff in a separate commit.
